File: code_indexer.py

import os
from typing import List, Dict, Any, Optional

import lancedb
import numpy as np
import pyarrow as pa
from langchain_openai import OpenAIEmbeddings
from langchain.text_splitter import RecursiveCharacterTextSplitter



class CodeIndexer:
    """
    CodeIndexer indexes code chunks into LanceDB, generating embeddings from detailed descriptions.
    It supports chunk splitting, embedding, and semantic/keyword search.
    """

    def __init__(self, db_path: str = "code_database", openai_api_key: Optional[str] = None):
        """
        Initialize the code indexer with LanceDB and OpenAI embeddings.
        """
        self.db_path = db_path
        self.db = lancedb.connect(db_path)

        if openai_api_key:
            os.environ["OPENAI_API_KEY"] = openai_api_key
        self.embeddings = OpenAIEmbeddings()

        self.text_splitter = RecursiveCharacterTextSplitter(
            chunk_size=1000, chunk_overlap=200, separators=["\n\n", "\n", " ", ""]
        )
        self.table = self._get_or_create_table()

    def _get_or_create_table(self):
        """
        Get existing table or create a new one with the required schema.
        """
        try:
            table = self.db.open_table("code_chunks")
            print("Using existing table: code_chunks")
            return table
        except Exception:
            print("Creating new table: code_chunks")
            schema = pa.schema([
                pa.field("id", pa.string()),
                pa.field("file_path", pa.string()),
                pa.field("file_name", pa.string()),
                pa.field("chunk_type", pa.string()),
                pa.field("chunk_name", pa.string()),
                pa.field("code", pa.string()),
                pa.field("description", pa.string()),
                pa.field("start_line", pa.int32()),
                pa.field("end_line", pa.int32()),
                pa.field("total_lines", pa.int32()),
                pa.field("file_size", pa.int32()),
                pa.field("embedding", pa.list_(pa.float32(), 1536)),
                pa.field("metadata", pa.string()),
            ])
            return self.db.create_table("code_chunks", schema=schema)


    def _create_chunk_id(self, file_path: str, chunk_type: str, chunk_name: str, start_line: int) -> str:
        """
        Create a unique ID for each chunk.
        """
        return f"{file_path}:{chunk_type}:{chunk_name}:{start_line}"


    def _split_long_code(self, code: str) -> List[str]:
        """
        Split long code into smaller chunks if needed.
        """
        if len(code) <= 1000:
            return [code]
        return self.text_splitter.split_text(code)


    def _get_embedding(self, text: str) -> List[float]:
        """
        Get embedding for text using OpenAI. Returns a zero vector on failure.
        """
        try:
            return self.embeddings.embed_query(text)
        except Exception as e:
            print(f"Error getting embedding: {e}")
            return [0.0] * 1536


    def _generate_description(self, chunk: Dict[str, Any], code_part: str) -> str:
        """
        Generate a detailed description for a code chunk, including file path, chunk info, and purpose.
        """
        desc = (
            f"File: {chunk['file_path']}\n"
            f"Chunk Name: {chunk.get('name', 'unnamed')}\n"
            f"Chunk Type: {chunk.get('type', '')}\n"
            f"Start Line: {chunk.get('start_line', '')}, End Line: {chunk.get('end_line', '')}\n"
            f"Total Lines: {chunk.get('total_lines', '')}\n"
            "Purpose: "
        )
        # Try to extract a docstring or comment from the code_part for purpose
        doc = ""
        for line in code_part.strip().splitlines():
            s = line.strip()
            if s.startswith('"""') or s.startswith("'''"):
                doc = s.strip('"\'')
                break
            if s.startswith('#'):
                doc = s.lstrip('#').strip()
                break
        if doc:
            desc += doc
        else:
            desc += "No explicit docstring or comment found. This chunk may define a function, class, or code block."
        return desc


    def index_chunks(self, chunks: List[Dict[str, Any]]) -> int:
        """
        Index code chunks into LanceDB with detailed description and embed the description instead of code.
        Returns the number of successfully indexed chunks.
        """
        indexed_count = 0
        for chunk in chunks:
            code_parts = self._split_long_code(chunk["code"])
            for i, code_part in enumerate(code_parts):
                chunk_id = self._create_chunk_id(
                    chunk["file_path"],
                    chunk["type"],
                    chunk.get("name", "unnamed"),
                    chunk["start_line"],
                )
                if len(code_parts) > 1:
                    chunk_id += f":part_{i}"
                description = self._generate_description(chunk, code_part)
                embedding = self._get_embedding(description)
                metadata = {
                    "node_type": chunk.get("node_type", ""),
                    "part_index": i,
                    "total_parts": len(code_parts),
                }
                record = {
                    "id": chunk_id,
                    "file_path": chunk["file_path"],
                    "file_name": chunk["file_name"],
                    "chunk_type": chunk["type"],
                    "chunk_name": chunk.get("name", "unnamed"),
                    "code": code_part,
                    "description": description,
                    "start_line": chunk["start_line"],
                    "end_line": chunk["end_line"],
                    "total_lines": chunk["total_lines"],
                    "file_size": chunk["file_size"],
                    "embedding": embedding,
                    "metadata": str(metadata),
                }
                try:
                    self.table.add([record])
                    indexed_count += 1
                except Exception as e:
                    print(f"Error indexing chunk {chunk_id}: {e}")
        print(f"Successfully indexed {indexed_count} chunks")
        return indexed_count


    def search_similar(self, query: str, limit: int = 10, threshold: float = 0.7) -> List[Dict[str, Any]]:
        """
        Search for similar code chunks using semantic search.
        Returns a list of result dicts with similarity scores.
        """
        query_embedding = self._get_embedding(query)
        results = (
            self.table.search(query_embedding, vector_column_name="embedding")
            .limit(limit)
            .to_pandas()
        )
        filtered_results = []
        for _, row in results.iterrows():
            similarity = self._cosine_similarity(query_embedding, row["embedding"])
            if similarity >= threshold:
                result_dict = row.to_dict()
                result_dict["similarity"] = similarity
                filtered_results.append(result_dict)
        filtered_results.sort(key=lambda x: x["similarity"], reverse=True)
        return filtered_results


    def _cosine_similarity(self, vec1: List[float], vec2: List[float]) -> float:
        """
        Calculate cosine similarity between two vectors.
        """
        vec1 = np.array(vec1)
        vec2 = np.array(vec2)
        norm1 = np.linalg.norm(vec1)
        norm2 = np.linalg.norm(vec2)
        if norm1 == 0 or norm2 == 0:
            return 0.0
        return float(np.dot(vec1, vec2) / (norm1 * norm2))


    def search_by_keyword(self, keyword: str, limit: int = 10) -> List[Dict[str, Any]]:
        """
        Search for code chunks containing specific keywords in code or chunk name.
        """
        results = (
            self.table.search(
                f"code LIKE '%{keyword}%' OR chunk_name LIKE '%{keyword}%'"
            )
            .limit(limit)
            .to_pandas()
        )
        return results.to_dict("records")


    def get_stats(self) -> Dict[str, Any]:
        """
        Get database statistics: total chunks, chunk type distribution, and file distribution.
        """
        total_chunks = len(self.table)
        df = self.table.to_pandas()
        chunk_types = df["chunk_type"].value_counts().to_dict()
        file_counts = df["file_name"].value_counts().to_dict()
        return {
            "total_chunks": total_chunks,
            "chunk_types": chunk_types,
            "files": file_counts,
        }

File: code-indexer.ts
// Note: Install these packages with npm:
// npm install @lancedb/lancedb @langchain/openai langchain @types/node

// For now, we'll use type declarations that match the expected API
// In a real implementation, you would install the actual packages

interface LanceDBConnection {
  openTable(name: string): Promise<LanceDBTable>;
  createTable(name: string, data: any[]): Promise<LanceDBTable>;
}

interface LanceDBTable {
  add(records: any[]): Promise<void>;
  vectorSearch(vector: number[]): {
    limit(n: number): {
      toArray(): Promise<any[]>;
    };
  };
  search(query: string): {
    limit(n: number): {
      toArray(): Promise<any[]>;
    };
  };
  toArray(): Promise<any[]>;
  delete(condition: string): Promise<void>;
}

interface OpenAIEmbeddingsInterface {
  embedQuery(text: string): Promise<number[]>;
}

interface TextSplitterInterface {
  splitText(text: string): Promise<string[]>;
}

// Mock implementations - replace with actual imports when packages are installed
const lancedb = {
  connect: async (path: string): Promise<LanceDBConnection> => {
    throw new Error("LanceDB not installed. Run: npm install @lancedb/lancedb");
  }
};

class OpenAIEmbeddings implements OpenAIEmbeddingsInterface {
  async embedQuery(text: string): Promise<number[]> {
    throw new Error("OpenAI embeddings not installed. Run: npm install @langchain/openai");
  }
}

class RecursiveCharacterTextSplitter implements TextSplitterInterface {
  constructor(options: { chunkSize: number; chunkOverlap: number; separators: string[] }) {
    // Mock constructor
  }
  
  async splitText(text: string): Promise<string[]> {
    throw new Error("LangChain text splitter not installed. Run: npm install langchain");
  }
}

// Type definitions
export interface CodeChunk {
  file_path: string;
  file_name: string;
  type: string;
  name?: string;
  code: string;
  start_line: number;
  end_line: number;
  total_lines: number;
  file_size: number;
  node_type?: string;
}

export interface CodeRecord {
  id: string;
  file_path: string;
  file_name: string;
  chunk_type: string;
  chunk_name: string;
  code: string;
  description: string;
  start_line: number;
  end_line: number;
  total_lines: number;
  file_size: number;
  embedding: number[];
  metadata: string;
}

export interface SearchResult extends CodeRecord {
  similarity: number;
}

interface Stats {
  total_chunks: number;
  chunk_types: Record<string, number>;
  files: Record<string, number>;
}

interface ChunkMetadata {
  node_type: string;
  part_index: number;
  total_parts: number;
}

/**
 * CodeIndexer indexes code chunks into LanceDB, generating embeddings from detailed descriptions.
 * It supports chunk splitting, embedding, and semantic/keyword search.
 */
export class CodeIndexer {
  private dbPath: string;
  private db!: LanceDBConnection;
  private embeddings: OpenAIEmbeddingsInterface;
  private textSplitter: TextSplitterInterface;
  private table!: LanceDBTable;

  constructor(dbPath: string = "code_database", openaiApiKey?: string) {
    this.dbPath = dbPath;
    
    if (openaiApiKey) {
      process.env.OPENAI_API_KEY = openaiApiKey;
    }
    
    this.embeddings = new OpenAIEmbeddings();
    this.textSplitter = new RecursiveCharacterTextSplitter({
      chunkSize: 1000,
      chunkOverlap: 200,
      separators: ["\n\n", "\n", " ", ""]
    });
  }

  /**
   * Initialize the database connection and table
   */
  async initialize(): Promise<void> {
    this.db = await lancedb.connect(this.dbPath);
    this.table = await this.getOrCreateTable();
  }

  /**
   * Get existing table or create a new one with the required schema.
   */
  private async getOrCreateTable(): Promise<LanceDBTable> {
    try {
      const table = await this.db.openTable("code_chunks");
      console.log("Using existing table: code_chunks");
      return table;
    } catch (error) {
      console.log("Creating new table: code_chunks");
      
      // Create sample data to infer schema
      const sampleData: CodeRecord[] = [{
        id: "sample",
        file_path: "sample.ts",
        file_name: "sample.ts",
        chunk_type: "function",
        chunk_name: "sample",
        code: "function sample() {}",
        description: "Sample function",
        start_line: 1,
        end_line: 1,
        total_lines: 1,
        file_size: 100,
        embedding: new Array(1536).fill(0),
        metadata: "{}"
      }];
      
      const table = await this.db.createTable("code_chunks", sampleData);
      // Delete the sample record
      await table.delete("id = 'sample'");
      return table;
    }
  }

  /**
   * Create a unique ID for each chunk.
   */
  private createChunkId(filePath: string, chunkType: string, chunkName: string, startLine: number): string {
    return `${filePath}:${chunkType}:${chunkName}:${startLine}`;
  }

  /**
   * Split long code into smaller chunks if needed.
   */
  private async splitLongCode(code: string): Promise<string[]> {
    if (code.length <= 1000) {
      return [code];
    }
    return await this.textSplitter.splitText(code);
  }

  /**
   * Get embedding for text using OpenAI. Returns a zero vector on failure.
   */
  private async getEmbedding(text: string): Promise<number[]> {
    try {
      return await this.embeddings.embedQuery(text);
    } catch (error) {
      console.error(`Error getting embedding: ${error}`);
      return new Array(1536).fill(0);
    }
  }

  /**
   * Generate a detailed description for a code chunk, including file path, chunk info, and purpose.
   */
  private generateDescription(chunk: CodeChunk, codePart: string): string {
    let desc = `File: ${chunk.file_path}\n` +
               `Chunk Name: ${chunk.name || 'unnamed'}\n` +
               `Chunk Type: ${chunk.type}\n` +
               `Start Line: ${chunk.start_line}, End Line: ${chunk.end_line}\n` +
               `Total Lines: ${chunk.total_lines}\n` +
               `Purpose: `;

    // Try to extract a docstring or comment from the code_part for purpose
    let doc = "";
    const lines = codePart.trim().split('\n');
    
    for (const line of lines) {
      const trimmed = line.trim();
      if (trimmed.startsWith('"""') || trimmed.startsWith("'''")) {
        doc = trimmed.replace(/['"]/g, '');
        break;
      }
      if (trimmed.startsWith('#') || trimmed.startsWith('//')) {
        doc = trimmed.replace(/^[#\/]+\s*/, '');
        break;
      }
      if (trimmed.startsWith('/**') || trimmed.startsWith('/*')) {
        doc = trimmed.replace(/^\/\*+\s*/, '').replace(/\*+\/$/, '');
        break;
      }
    }

    if (doc) {
      desc += doc;
    } else {
      desc += "No explicit docstring or comment found. This chunk may define a function, class, or code block.";
    }

    return desc;
  }

  /**
   * Index code chunks into LanceDB with detailed description and embed the description instead of code.
   * Returns the number of successfully indexed chunks.
   */
  async indexChunks(chunks: CodeChunk[]): Promise<number> {
    let indexedCount = 0;
    
    for (const chunk of chunks) {
      const codeParts = await this.splitLongCode(chunk.code);
      
      for (let i = 0; i < codeParts.length; i++) {
        const codePart = codeParts[i];
        let chunkId = this.createChunkId(
          chunk.file_path,
          chunk.type,
          chunk.name || "unnamed",
          chunk.start_line
        );
        
        if (codeParts.length > 1) {
          chunkId += `:part_${i}`;
        }
        
        const description = this.generateDescription(chunk, codePart);
        const embedding = await this.getEmbedding(description);
        
        const metadata: ChunkMetadata = {
          node_type: chunk.node_type || "",
          part_index: i,
          total_parts: codeParts.length
        };
        
        const record: CodeRecord = {
          id: chunkId,
          file_path: chunk.file_path,
          file_name: chunk.file_name,
          chunk_type: chunk.type,
          chunk_name: chunk.name || "unnamed",
          code: codePart,
          description: description,
          start_line: chunk.start_line,
          end_line: chunk.end_line,
          total_lines: chunk.total_lines,
          file_size: chunk.file_size,
          embedding: embedding,
          metadata: JSON.stringify(metadata)
        };
        
        try {
          await this.table.add([record]);
          indexedCount++;
        } catch (error) {
          console.error(`Error indexing chunk ${chunkId}: ${error}`);
        }
      }
    }
    
    console.log(`Successfully indexed ${indexedCount} chunks`);
    return indexedCount;
  }

  /**
   * Search for similar code chunks using semantic search.
   * Returns a list of result objects with similarity scores.
   */
  async searchSimilar(query: string, limit: number = 10, threshold: number = 0.7): Promise<SearchResult[]> {
    const queryEmbedding = await this.getEmbedding(query);
    
    const results = await this.table
      .vectorSearch(queryEmbedding)
      .limit(limit)
      .toArray();
    
    const filteredResults: SearchResult[] = [];
    
    for (const row of results) {
      const similarity = this.cosineSimilarity(queryEmbedding, row.embedding);
      if (similarity >= threshold) {
        const result: SearchResult = {
          ...row,
          similarity: similarity
        };
        filteredResults.push(result);
      }
    }
    
    filteredResults.sort((a, b) => b.similarity - a.similarity);
    return filteredResults;
  }

  /**
   * Calculate cosine similarity between two vectors.
   */
  private cosineSimilarity(vec1: number[], vec2: number[]): number {
    if (vec1.length !== vec2.length) {
      return 0.0;
    }
    
    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;
    
    for (let i = 0; i < vec1.length; i++) {
      dotProduct += vec1[i] * vec2[i];
      norm1 += vec1[i] * vec1[i];
      norm2 += vec2[i] * vec2[i];
    }
    
    norm1 = Math.sqrt(norm1);
    norm2 = Math.sqrt(norm2);
    
    if (norm1 === 0 || norm2 === 0) {
      return 0.0;
    }
    
    return dotProduct / (norm1 * norm2);
  }

  /**
   * Search for code chunks containing specific keywords in code or chunk name.
   */
  async searchByKeyword(keyword: string, limit: number = 10): Promise<CodeRecord[]> {
    const results = await this.table
      .search(`code LIKE '%${keyword}%' OR chunk_name LIKE '%${keyword}%'`)
      .limit(limit)
      .toArray();
    
    return results;
  }

  /**
   * Get database statistics: total chunks, chunk type distribution, and file distribution.
   */
  async getStats(): Promise<Stats> {
    const allRecords = await this.table.toArray();
    const totalChunks = allRecords.length;
    
    const chunkTypes: Record<string, number> = {};
    const files: Record<string, number> = {};
    
    for (const record of allRecords) {
      // Count chunk types
      chunkTypes[record.chunk_type] = (chunkTypes[record.chunk_type] || 0) + 1;
      
      // Count files
      files[record.file_name] = (files[record.file_name] || 0) + 1;
    }
    
    return {
      total_chunks: totalChunks,
      chunk_types: chunkTypes,
      files: files
    };
  }

  /**
   * Close the database connection
   */
  async close(): Promise<void> {
    // LanceDB connections are automatically managed
    // This method is provided for consistency with the Python version
  }
}
